File: src/world_state.rs
use crate::model::world::WorldState;

use crate::PgPool;
use anyhow::Context;
use slog::{debug, error, Logger};
use tokio::time::{self, Duration};

#[derive(Clone)]
pub struct SharedState(pub left_right::ReadHandle<WorldState>);

unsafe impl Send for SharedState {}
unsafe impl Sync for SharedState {}

pub struct StateWriter(pub left_right::WriteHandle<WorldState, WorldStateOps>);

pub enum WorldStateOps {
    Set(WorldState),
}

impl left_right::Absorb<WorldStateOps> for WorldState {
    fn absorb_first(&mut self, operation: &mut WorldStateOps, _other: &Self) {
        match operation {
            WorldStateOps::Set(w) => self.clone_from(w),
        }
    }

    fn absorb_second(&mut self, operation: WorldStateOps, _other: &Self) {
        match operation {
            WorldStateOps::Set(w) => *self = w,
        }
    }
}

pub fn init_state() -> (StateWriter, SharedState) {
    let (w, r) = left_right::new::<WorldState, WorldStateOps>();
    (StateWriter(w), SharedState(r))
}

pub async fn refresh_state_job(
    pool: PgPool,
    logger: Logger,
    mut state: StateWriter,
    interval: Duration,
) -> anyhow::Result<()> {
    let mut interval = time::interval(interval);

    loop {
        interval.tick().await;
        debug!(logger, "Reading world state");

        let new_state = load_state(pool.clone(), logger.clone()).await?;
        state.0.append(WorldStateOps::Set(new_state));
        state.0.publish();

        debug!(logger, "Reading world state - done");
    }
}

// TODO queen tag?
pub async fn load_state(pool: PgPool, logger: Logger) -> anyhow::Result<WorldState> {
    struct Foo {
        world_time: i64,
        payload: serde_json::Value,
    }

    sqlx::query_as!(
        Foo,
        r#"
        SELECT t.world_time, t.payload
        FROM world_output t
        ORDER BY t.created DESC
        LIMIT 1
        "#
    )
    .fetch_one(&pool)
    .await
    .map(
        |Foo {
             world_time,
             payload,
         }| WorldState {
            time: world_time,
            payload,
        },
    )
    .map_err(|err| {
        error!(logger, "Failed to read world state {:?}", err);
        err
    })
    .with_context(|| "Failed to read world state")
}

File: world_state/world_state.go
package world_state

import (
	"encoding/json"
	"github.com/jmoiron/sqlx"
)

type WorldState struct {
	Bots           map[string]interface{} `json:"bots"`
	Structures     map[string]interface{} `json:"structures"`
	Resources      map[string]interface{} `json:"resources"`
	Rooms          map[string]interface{} `json:"rooms"`
	Users          interface{}            `json:"users"`
	Terrain        interface{}            `json:"terrain"`
	RoomProperties interface{}            `json:"roomProperties"`
	GameConfig     GameConfig             `json:"gameConfig"`
	Time           int64
}

type GameConfig struct {
	ExecutionLimit int `json:"execution_limit"`
}

var WorldStateQuery = `
SELECT t.payload, t.world_time
FROM world_output t
ORDER BY t.created DESC
LIMIT 1
`

func GetLatestWorldState(db *sqlx.DB) (*WorldState, error) {

	type WorldQResult struct {
		Payload []byte `db:"payload"`
		Time    int64  `db:"world_time"`
	}

	results := []WorldQResult{}
	err := db.Select(&results, WorldStateQuery)
	if err != nil {
		return nil, err
	}
	var state WorldState
	err = json.Unmarshal(results[0].Payload, &state)
	if err != nil {
		return nil, err
	}
	state.Time = results[0].Time
	return &state, nil
}
